### DISCUSSION TITLE
How does the Websocket server proactively send messages to clients?

### DISCUSSION THREAD

**Discussion by qilivip** [actix/actix-web#2729] [Q&A]
How does the Websocket server proactively send messages to clients?
Thanks!!!

**Comment by qilivip** [comment]
Problem solved!
https://github.com/actix/examples/blob/master/websockets/chat

server.rs
```
// add 
/// Send message to specific room
#[derive(Message)]
#[rtype(result = "()")]
pub struct ServerMessage {
    /// skip_id of the client session
    pub skip_id: usize,
    /// Peer message
    pub msg: String,
    /// Room name
    pub room: String,
}

/// Handler for Message message.
impl Handler<ServerMessage> for ChatServer {
    type Result = ();

    fn handle(&mut self, msg: ServerMessage, _: &mut Context<Self>) {
        self.send_message(&msg.room, msg.msg.as_str(), msg.skip_id);
    }
}
```
main.rs
```
// add
let thread_chat_server = chat_server.clone();

thread::spawn(move || {
    loop {
        // chat_server1.send_message("main", "hello world", 0);
        let msg = ServerMessage{ skip_id: 0, msg: "hello world".to_string(), room: "Main".to_string() };
        let _ = thread_chat_server.try_send(msg);
        thread::sleep(Duration::from_millis(1000));
    }
});
```

**Reply by fakeshadow** [reply]
Glad to see you figure out your problem but I'm afraid your code is not optimal. Do you mind if I unmark your answer as suggested one as it can be misleading to others new to Rust.

For proper way of doing interval task the example already showed it in websocket heartbeat handling https://github.com/actix/examples/blob/453311c296c05e263b95055178468d74dcc6b209/websockets/chat/src/session.rs#L38
`Context::run_interval` is how you run interval task with an actor.
